File: frontend/src/api/paper.ts
import { api } from "./client";

export type Position = {
  symbol: string;
  quantity: number;
  avg_cost: number;
  last_price: number | null;
  prev_close: number | null;
  market_value: number | null;
  unrealized_pnl: number | null;
  unrealized_pnl_pct: number | null;
  day_change: number | null;
  day_change_pct: number | null;
};

export type PendingSettlement = {
  amount: number;
  settles_at: string;
};

export type MarketStatus = {
  open: boolean;
  session: "pre" | "regular" | "post" | "closed";
  next_open_iso: string;
  now_iso: string;
};

export type Portfolio = {
  cash: number;
  buying_power: number;
  pending_settlement: number;
  reserved_for_orders: number;
  pending_settlements: PendingSettlement[];
  positions: Position[];
  market_value: number;
  total_value: number;
  initial_cash: number;
  realized_pnl: number;
  market_status: MarketStatus;
};

export type Trade = {
  id: number;
  symbol: string;
  side: "buy" | "sell";
  quantity: number;
  price: number;
  realized_pnl: number | null;
  executed_at: string;
};

export type LimitOrder = {
  id: number;
  symbol: string;
  side: "buy" | "sell";
  quantity: number;
  limit_price: number;
  status: string;
  created_at: string;
  filled_at: string | null;
};

export const paperApi = {
  portfolio: () => api.get<Portfolio>("/api/paper/portfolio"),
  trades: () => api.get<Trade[]>("/api/paper/trades"),
  buy: (symbol: string, quantity: number) =>
    api.post<Trade>("/api/paper/buy", { symbol, quantity }),
  sell: (symbol: string, quantity: number) =>
    api.post<Trade>("/api/paper/sell", { symbol, quantity }),
  reset: () => api.post<Portfolio>("/api/paper/reset", {}),
  marketStatus: () => api.get<MarketStatus>("/api/paper/market-status"),
  listOrders: () => api.get<LimitOrder[]>("/api/paper/orders"),
  placeOrder: (params: {
    symbol: string;
    side: "buy" | "sell";
    quantity: number;
    limit_price: number;
  }) => api.post<LimitOrder>("/api/paper/orders", params),
  cancelOrder: (id: number) => api.del<LimitOrder>(`/api/paper/orders/${id}`),
};

File: backend/app/schemas/paper.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class TradeIn(BaseModel):
    symbol: str = Field(min_length=1, max_length=16)
    quantity: int = Field(gt=0)


class PositionOut(BaseModel):
    symbol: str
    quantity: int
    avg_cost: float
    last_price: float | None
    prev_close: float | None = None
    market_value: float | None
    unrealized_pnl: float | None
    unrealized_pnl_pct: float | None
    day_change: float | None = None
    day_change_pct: float | None = None

    model_config = {"from_attributes": True}


class TradeOut(BaseModel):
    id: int
    symbol: str
    side: Literal["buy", "sell"]
    quantity: int
    price: float
    realized_pnl: float | None = None
    executed_at: datetime

    model_config = {"from_attributes": True}


class PendingSettlementOut(BaseModel):
    amount: float
    settles_at: str


class MarketStatusOut(BaseModel):
    open: bool
    session: Literal["pre", "regular", "post", "closed"]
    next_open_iso: str
    now_iso: str


class PortfolioOut(BaseModel):
    cash: float
    buying_power: float
    pending_settlement: float
    reserved_for_orders: float = 0.0
    pending_settlements: list[PendingSettlementOut] = []
    positions: list[PositionOut]
    market_value: float
    total_value: float
    initial_cash: float = 100000.0
    realized_pnl: float = 0.0
    market_status: MarketStatusOut


class LimitOrderIn(BaseModel):
    symbol: str = Field(min_length=1, max_length=16)
    side: Literal["buy", "sell"]
    quantity: int = Field(gt=0)
    limit_price: float = Field(gt=0)


class LimitOrderOut(BaseModel):
    id: int
    symbol: str
    side: Literal["buy", "sell"]
    quantity: int
    limit_price: float
    status: str
    created_at: datetime
    filled_at: datetime | None

    model_config = {"from_attributes": True}
